### PR TITLE
[Spark-27406][SQL]UnsafeArrayData serialization breaks when two machines have different Oops size

### PR DESCRIPTION
**Pull request by pengbo**
## What changes were proposed in this pull request?
ApproxCountDistinctForIntervals holds the UnsafeArrayData data to initialize endpoints. When the UnsafeArrayData is serialized with Java serialization, the BYTE_ARRAY_OFFSET in memory can change if two machines have different pointer width (Oops in JVM).

This PR fixes this issue by using the same way in https://github.com/apache/spark/pull/9030

## How was this patch tested?
Manual test has been done in our tpcds environment and regarding unit test case has been added as well

### CODE REVIEW

**Comment by pengbo** [comment]
@rxin Can you please have a look?

**Comment by sandeep-katta** [comment]
Same thing is required to handle for UnsafeMapData

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java [hunk 1]
```diff
@@ -52,7 +58,7 @@
  * Instances of `UnsafeArrayData` act as pointers to row data stored in this format.
  */
 
-public final class UnsafeArrayData extends ArrayData {
+public final class UnsafeArrayData extends ArrayData implements Externalizable {
```
Comment (sandeep-katta) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeArrayData.java]: implement KryoSerializable also

**Comment by sandeep-katta** [comment]
cc @cloud-fan , @srowen  @HyukjinKwon @rxin

**Comment by cloud-fan** [comment]
can we create a trait for all these 3 unsafe data structures (row, array, map)?

**Comment by pengbo** [comment]
@cloud-fan sounds reasonable, I will do it by this PR.

**Comment by HeartSaVioR** [comment]
Maybe reconsidering #24050 or picking up the approach #24050 and minimize the boundary of area (like restricting to `getBytes()`)?

**Comment by cloud-fan** [comment]
ok to test

**Comment by cloud-fan** [comment]
let's merge it first and think about code refactor later.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by cloud-fan** [comment]
can you send another PR to 2.4? it conflicts. Thanks!

**Comment by pengbo** [comment]
Okay, It will be done tonight.
I will create another PR for UnsafeMapData and refactoring later this week.

**Pull request closed without merging**